pkg/ddl/partition.go

```diff
@@ -2428,236 +2357,245 @@ func removeTiFlashAvailablePartitionIDs(tblInfo *model.TableInfo, pids []int64)
-			tblInfo,
-			&model.PartitionInfo{Definitions: newPartitions},
-			&model.PartitionInfo{Definitions: oldPartitions},
-		)
-		err = asyncNotifyEvent(jobCtx, truncatePartitionEvent, job, noSubJob, w.sess)
-		if err != nil {
-			return ver, errors.Trace(err)
-		}
-
-		// Finish this job.
-		job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
-		// A background job will be created to delete old partition data.
-		job.FillFinishedArgs(&model.TruncateTableArgs{
-			OldPartitionIDs: oldIDs,
-		})
-
-		return ver, err
+		return ver, errors.Trace(dbterror.ErrPartitionMgmtOnNonpartitioned)
 	}
 
-	// When table has global index, public->deleteOnly->deleteReorg->none schema changes should be handled.
+	var oldDefinitions []model.PartitionDefinition
+	var newDefinitions []model.PartitionDefinition
+
 	switch job.SchemaState {
 	case model.StatePublic:
-		// Step1: generate new partition ids
-		truncatingDefinitions := make([]model.PartitionDefinition, 0, len(oldIDs))
-		for i, oldID := range oldIDs {
-			for j := 0; j < len(pi.Definitions); j++ {
-				def := &pi.Definitions[j]
-				if def.ID == oldID {
-					truncatingDefinitions = append(truncatingDefinitions, def.Clone())
-					def.ID = newIDs[i]
-					break
-				}
-			}
-		}
-		pi.DroppingDefinitions = truncatingDefinitions
+		// This work as a flag to ignore Global Index entries from the new partitions!
 		pi.NewPartitionIDs = newIDs[:]
+		pi.DDLAction = model.ActionTruncateTablePartition
 
+		job.SchemaState = model.StateWriteOnly
+		pi.DDLState = job.SchemaState
+		return updateVersionAndTableInfo(jobCtx, job, tblInfo, true)
+	case model.StateWriteOnly:
+		oldDefinitions, newDefinitions, err = replaceTruncatePartitions(job, jobCtx.metaMut, tblInfo, oldIDs, newIDs)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		preSplitAndScatter(w.sess.Context, jobCtx.store, tblInfo, newDefinitions)
+		// This work as a flag to ignore Global Index entries from the old partitions!
+		pi.DroppingDefinitions = oldDefinitions
+		// And we don't need to filter for new partitions any longer
 		job.SchemaState = model.StateDeleteOnly
 		pi.DDLState = job.SchemaState
-		pi.DDLAction = job.Type
-		ver, err = updateVersionAndTableInfo(jobCtx, job, tblInfo, true)
+		return updateVersionAndTableInfo(jobCtx, job, tblInfo, true)
 	case model.StateDeleteOnly:
-		// This state is not a real 'DeleteOnly' state, because tidb does not maintaining the state check in partitionDefinition.
-		// Insert this state to confirm all servers can not see the old partitions when reorg is running,
-		// so that no new data will be inserted into old partitions when reorganizing.
+		// Now we don't see the old partitions, but other sessions may still use them.
+		// So to keep the Global Index consistent, we will still keep it up-to-date with
+		// the old partitions, as well as the new partitions.
+		// Also ensures that no writes will happen after GC in DeleteRanges.
+
```

Comment:
I don't understand why this intermediate state is needed. When L2544 is executed, other TiDBs must have switched to the StateDeleteOnly state. This state seems do nothing here.

Reply:
No, when L2544 is executed, it only means that the current go routine has updated the state/other TiDBs must have switched to the `StateWriteOnly` state. The tableInfo will be written to the KV store and either patched by ApplyDiff or fully reloaded, which will happen asyncronously between different TiDB servers (domains). The goroutine executing this DDL may wait until there are no transactions using the previous schema version (see how MDL is implemented).

But it still means that as soon as updateVersionAndTableInfo + commit in calling function is done, then some TiDB servers can see the new SchemaVersion and some will be using the 'old' SchemaVersion.
And in that case if `StateDeleteOnly` was not used, then `StateWriteOnly` could write to the old set of partitions during `StateDeleteReorganization`, meaning that depending on timing, it may also write to the Global Index, after DeleteReorganization has already cleaned that record range, leaving an orphan entry in the Global index.

Reply:
We have `waitVersionSynced` function which will wait all TiDB schemas are synced. So when  `cleanGlobalIndexEntriesFromDroppedPartitions` execute, all TiDB nodes should in `model.StateDeleteReorganization` state.

Reply:
Ah, thank you for the correction!
After trying to remove the StateDeleteOnly mode, I realized that there is one more reason to keep it:
If going directly from StateWriteOnly (where it would not read the new partitions, but still get duplicate key errors on the Global Index) to StateDeleteReorganization (where it would not read the old partitions, and directly overwrite Global Index entries pointing to old partitions) there would be an anomaly like:
StateWriteOnly session 'wo> '
StateDeleteReorganization session 'dr> '
```mysql
init> create table t (a int primary key, b varchar(255), unique index (`b`) global) partition by hash (a) partitions 2;
wo> insert into t values (1,1,"WO"); -- success
dr> insert into t values (1,1,"DR"); -- success
wo> select * from t where a = 1; -- 1 1 WO
dr> select * from t where a = 1; -- 1 1 DR
```
Regardless of global index or not, the above would happen (with current PR implementation).
We could also remove the optimization for overwriting old global index entries in StateDeleteReorganization, which would then fail the 'dr> ' insert due to duplicated global index entry. But that state/error would potentially exists for a long time, if the truncated partition had a lot of rows.

@Defined2014 I would still propose to keep the StateDeleteOnly where the three states would mean:
- StateWriteOnly
  - Aware of new partitions, so it would filter out those partitions from Global Indexes on read, and fail on duplicated key on write.
- StateDeleteOnly
  - Using the new (initally empty) partitions, but will give duplicate errors on the Global Index if writes collide with entries from old (dropped) partitions, since those entries may still be used by sessions in StateWriteOnly.
  - This state will only take as long as it takes for updating the Schema Version across a cluster, so limited time not depending on how many rows the dropped partitions has.
- StateDeleteReorganization
  - Using the new partitions, and if there is a duplicate Global Index entry for an old partition, it will simply overwrite it, since those entries are no-longer used.

What do you think?
  - During this state all Global Index entries pointing to old partitions will be deleted by a reorg DDL worker.

Reply:
There is another small question. If the `StateDeleteOnly` state is used, will the above situation still occur for the unique key of non-global index?

Reply:
Yes, non-global indexes (local indexes) will be able to have 'non-visible' duplicates, since two sessions can have different sets of partitions between `StateWriteOnly` and `StateDeleteOnly`. But it is not really related to the specific state, just that the transition between having the old set of partitions visible, to have the new set of partitions visible. A less user friendly way would be to either have a state where the truncated partitions are completely blocked or check both old and new partitions on write (which is kind of what Global Index does).